console-subscriber/tests/support/subscriber.rs

```diff
@@ -0,0 +1,318 @@
+                .await;
+        });
+    });
+
+    let actual_tasks = join_handle
+        .join()
+        .expect("console-test error: failed to join 'console-subscriber' thread");
+
+    if let Err(test_failure) = validate_expected_tasks(expected_tasks, actual_tasks) {
+        panic!("Test failed: {test_failure}")
+    }
+}
+
+/// Starts the console server.
+///
+/// The server will start serving over its side of the duplex stream.
+///
+/// Once the server gets spawned into its task, the test state is advanced
+/// to the `ServerStarted` step. This function will then wait until the test
+/// state reaches the `UpdatesRecorded` step (indicating that all validation of the
+/// received updates has been completed) before dropping the aggregator.
+///
+/// # Test State
+///
+/// 1. Advances to: `ServerStarted`
+/// 2. Waits for: `UpdatesRecorded`
+async fn console_server(
+    server: console_subscriber::Server,
+    server_stream: DuplexStream,
+    mut test_state: TestState,
+) {
+    let ServerParts {
+        instrument_server: service,
+        aggregator,
+        ..
+    } = server.into_parts();
+    let aggregate = task::Builder::new()
+        .name("console::aggregate")
+        .spawn(aggregator.run())
+        .expect("client-console error: couldn't spawn aggregator");
+    Server::builder()
+        .add_service(service)
+        .serve_with_incoming(futures::stream::iter(vec![Ok::<_, std::io::Error>(
+            server_stream,
+        )]))
+        .await
+        .expect("client-console error: couldn't start instrument server.");
+    test_state.advance_to_step(TestStep::ServerStarted);
+
+    test_state.wait_for_step(TestStep::UpdatesRecorded).await;
+    aggregate.abort();
+}
+
+/// Starts the console client and validates the expected tasks.
+///
+/// First we wait until the server has started (test step `ServerStarted`), then
+/// the client is connected to its half of the duplex stream and we start recording
+/// the actual tasks.
+///
+/// Once recording finishes (see [`record_actual_tasks()`] for details on the test
+/// state condition), the actual tasks returned.
+///
+/// # Test State
+///
+/// 1. Waits for: `ServerStarted`
+/// 2. Advances to: `ClientConnected`
+async fn console_client(client_stream: DuplexStream, mut test_state: TestState) -> Vec<ActualTask> {
+    test_state.wait_for_step(TestStep::ServerStarted).await;
+
+    let mut client_stream = Some(client_stream);
+    let channel = Endpoint::try_from("http://[::]:6669")
+        .expect("Could not create endpoint")
+        .connect_with_connector(service_fn(move |_: Uri| {
+            let client = client_stream.take();
+
+            async move {
+                if let Some(client) = client {
+                    Ok(client)
+                } else {
+                    Err(std::io::Error::new(
+                        std::io::ErrorKind::Other,
+                        "Client already taken",
+                    ))
+                }
```

Comment:
nit: isn't this just `Option::ok_or_else`?
```suggestion
                client.ok_or_else(|| std::io::Error::new(std::io::ErrorKind::Other, "Client already taken"))
```